helix-core/src/main/java/org/apache/helix/controller/rebalancer/constraint/ExcessiveTopStateResolver.java

```diff
@@ -0,0 +1,112 @@
+ */
+public class ExcessiveTopStateResolver implements AbnormalStateResolver {
+  private static final Logger LOG = LoggerFactory.getLogger(ExcessiveTopStateResolver.class);
+
+  /**
+   * The current states are not valid if there are more than 2 top state replicas for a single top
+   * state state model.
+   */
+  @Override
+  public boolean isCurrentStatesValid(final CurrentStateOutput currentStateOutput,
+      final String resourceName, final Partition partition, StateModelDefinition stateModelDef) {
+    if (!stateModelDef.isSingleTopStateModel()) {
+      return true;
+    }
+    if (currentStateOutput.getCurrentStateMap(resourceName, partition).values().stream()
+        .filter(state -> state.equals(stateModelDef.getTopState())).count() > 1) {
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  public Map<String, String> computeRecoveryAssignment(final CurrentStateOutput currentStateOutput,
+      final String resourceName, final Partition partition, StateModelDefinition stateModelDef,
+      List<String> preferenceList) {
+    Map<String, String> currentStateMap =
+        currentStateOutput.getCurrentStateMap(resourceName, partition);
+    if (isCurrentStatesValid(currentStateOutput, resourceName, partition, stateModelDef)) {
+      // This method should not be triggered when the mapping is valid.
+      // Log the warning for debug purposes.
+      LOG.warn("The input current state map {} is valid, return the original current state.",
+          currentStateMap);
+      return currentStateMap;
+    }
+
+    Map<String, String> recoverMap = new HashMap<>(currentStateMap);
+    String recoveryState = stateModelDef
+        .getNextStateForTransition(stateModelDef.getTopState(), stateModelDef.getInitialState());
+
+    // 1. We have to reset the expected top state replica host if it is hosting the top state
+    // replica. Otherwise, the potential data issue will never be fixed there.
+    if (preferenceList != null && !preferenceList.isEmpty()) {
+      String expectedTopStateHost = preferenceList.get(0);
+      if (recoverMap.get(expectedTopStateHost).equals(stateModelDef.getTopState())) {
+        recoverMap.put(expectedTopStateHost, recoveryState);
+      }
+    }
+
+    // 2. To minimize the impact of the resolution, we want to reserve one top state replica even
+    // during the recovery process.
+    boolean hasReservedTheTopState = false;
+    for (String instance : recoverMap.keySet()) {
+      if (recoverMap.get(instance).equals(stateModelDef.getTopState())) {
+        if (hasReservedTheTopState) {
+          recoverMap.put(instance, recoveryState);
+        } else {
+          hasReservedTheTopState = true;
+        }
+      }
+    }
+    // Here's what we expect to happen next:
+    // 1. The partition assignment is changed to the proposed recovery state. Or it may be halfway
+    // there.
+    // 2. If the new current state is still invalid, then continue fixing it with the same logic.
```

Comment:
You mean the rebalancer will continue to run this for all N resources until everything has been resolved?

Or will it check for each resource, say, resource N, I check isValid() and correctCurrentStates() until isValid() returns true?

What I am not too unsure about is what you said about "new current state" being "still invalid". Is it possible that the result is invalid after correction?

Reply:
There is no other resource considered here. Let me add "then the resolver continues fixing it with the same logic."

Reply:
> What I am not too unsure about is what you said about "new current state" being "still invalid". Is it possible that the result is invalid after correction?

I think that's why you have the question why the method is named computeRecoveryAssignment instead of computeCorrectedAssignment. The result is push the rebalancer to move one step toward the fix. But it is not necessarily fixing the states directly. Like in this case, it cannot be done within one step without impact the availability. This is the first point.

Secondly, even for some resolvers, they fix the things in one step, there is no guarantee that the participant will finish the state transition immediately.